Add Debugger-like visualization to variables

henriquelalves opened the pull request:
This Pull Request adds:
- A tooltip for variables showing their value while the code is running.
- A "debugger" visualization in which users can check the script variables values while the example runs.

Xananax commented:
It's very cool! Impressive work.

Do I understand correctly that the example implementation is in "Chapter 18 - introduction to For Loops", scene `res://course/lesson-18-for-loops/visuals/PrintForLoop.tscn`?

If yes, it seems the hover popups don't show in situ (they show when I run the scene with F6).

Apart from that, I have some comments regarding the code.

Currently, if I understand properly, for the debugger to work, you need two dependent, but independently entered pieces of information:

- export a `monitored_variables` array
- add a `RunnableCodeExampleDebugger` node

If that is correct, I would suggest making it so it's not possible to have one without the other, to minimize both necessary documentation, and possible mistakes. For example, we could:

1. Add the `RunnableCodeExampleDebugger` node at runtime, if `monitored_variables` is found
2. Add the `monitored_variables` export to `RunnableCodeExampleDebugger` and read it from there

Of these 2, I think the second is preferable, because of its greater clarity and discoverability. It also allows to reshape the debugger's size or theme if necessary.

henriquelalves commented:
Completely agree with you there. I don't like the first refactor option because to add the DebuggerScene, I have to replace the root of the example scene to a VBoxContainer, and making this change may be really cumbersome.

I refactored using your second idea. The only problem is that we can only highlight variables when the DebuggerScene is present, but I don't know if we ever want to highlight variables and don't add a debugger scene anyway.

I'm also adding the Debugger scene on lesson 18 examples to test it, but I can remove it before merging. Weird thing is that I'm able to see it on lesson 18 though, can you confirm the error of the examples not having the debugger scene on the lesson 18?

Xananax commented:
Regarding the pop-up: I do see the _debugger_, but I couldn't see the _pop-up number_ on hover.

Trying again, it appears the issue is if I press `run()` before pressing `step()`. The highlights and numbers then stop appearing. If I press only `step`, then all is good.

Apart from that, I got nothing, it's looking good. I didn't do a full code review, once the bug is fixed, please add me as a reviewer or ping me on discord, I'll go through the code with more scrutiny.

On another hand: Should implementation sites be part of this PR, or is lesson 18 the only implementation site?

Once the bug is fixed, I suggest using the debugger, in this same PR, in all the lessons where it can/should be used, as to uncover more bugs and use cases before merging.

If you decide to do that, then we can review the code afterward (it might change).

henriquelalves commented:
Just updated all relevant lessons! My priority was to change lessons with more than one or two lines of code, in which a variable changes value. I also fixed one or two issues along the way, so now all lessons should be working.

henriquelalves commented:
Tagging related content issues this PR may have affected/solved.

- Lesson 8
  - #659 
- Lesson 15
  - #639 
  - #574 
- Lesson 17
  - #636

Xananax reviewed (commented):
Excellent! This is really neat.

Apart from a few changes to the types, I didn't have anything to review.

A few comments/questions

## 1 - Bug Report

Currently, pressing "reset" doesn't seem to reset the value, at least in that exercise

![debugger](https://user-images.githubusercontent.com/836082/191278623-e4220291-308a-44e5-8878-9bef45dc0f5c.png)

I did not check the others

## 2 - Documentation

I can guess the process to add the debugger and the arrow highlighter, but I'd like a formal description of the process which we can give in general to people who want to use it.

A few bullet points is enough, it doesn't need to be detailed or in-depth, it's for internal usage

## 3 - Automation

For the purposes of the current project, I think the scope of this is very good, but it seems to me like the process is involved and has several points where errors can easily happen (I believe that's the reason for the bug described above). For a potential future update, do you envision any way to automate completely or partially the process, to reduce potential errors and simplify workflow?

Alternatively or additionally, is there any way to error check better that everything is set up correctly?

henriquelalves commented:
# 1 - Bug Report
Good catch! Didn't notice there was a bug with reset(). I had to fix it on all examples by overriding the "reset()" function and setting the variable to their initial value - it's one more thing to remember when adapting a code example, but I don't think there is any other way to do this without making it way less clear how the debugger interacts with the code.

# 2 - Documentation
Just created this small gist with some step-by-step instructions, I can update it if there is any questions on how to change an example: https://gist.github.com/henriquelalves/365f46893e99d8b6cacfeb37d4901399

# 3 - Automation
The steps to adapt an example to add a Debugger is somewhat straightforward, but I think we could automate it with an addon to make it simpler (changing the root of the example to a VBox and addding a Debugger scene). I think what is error-prone is adapting the examples to run in step-by-step mode, but I don't know if there is an easy way to automate it without making the whole process much less clear on how it works.

NathanLovato commented:
When testing or pushing a PR, be sure to test every single change in the context the user will see it. I.e. if you tested individual examples by pressing F6, don't do that. In L15 for instance, the traffic light demo is cut off:

![image](https://user-images.githubusercontent.com/12694995/191437425-29676367-399e-4680-94a9-c205ee8575da.png)

In lesson 13 there are these visual artifacts

![image](https://user-images.githubusercontent.com/12694995/191437550-158b2fbe-0dd8-4c9f-9b14-43e1c26f16aa.png)

I spotted a couple of little things like these re-reading the full lessons, this still lacks some cleanup

NathanLovato commented:
Getting a weird issue: the variable highlight breaks in Godot 3.5, but not always, and displays like this.

![image](https://user-images.githubusercontent.com/12694995/191453440-2ce91177-ac93-4d89-b32a-e1edd438c9a5.png)

henriquelalves commented:
Just added an "UNINTIALIZED_VARIABLE_VALUE" constant  on the Debugger, so I'm hiding the variable if it's the unintialized string. I also updated the `reset()` function to always be called once before running (we were already doing that but implicitly, since we were running "clear_messages()" when the function started).

NathanLovato commented on ui/components/RunnableCodeExample.gd:
```diff
@@ -313,11 +326,12 @@ func _set_scene_instance(new_scene_instance: CanvasItem) -> void:
 				rect.position += offset
 				rect.size.x = (rect.size.x * variable_name.length()) + 4
 
-				var monitored_variable: CodeExampleVariableUnderline = CodeExampleVariableUnderlineScene.instance()
```
I invite you to use the formatter gdformat. Without space is the official gdscript style guide

NathanLovato commented on ui/components/RunnableCodeExample.gd:
```diff
@@ -267,26 +270,36 @@ func _set_scene_instance(new_scene_instance: CanvasItem) -> void:
 
 	# Create widgets that underline a variable and display a variable's value
 	# when hovering with the mouse.
-	var monitored_variables := debugger.monitored_variables
```
You're getting an error with type inference here? it's working fine on my end
